allow custom encoding/decoding of postgres' NULL on a per query basis

talentdeficit opened the pull request:
@fishcakez 

this still needs to be documented and the opt name is probably not ideal but making this pr for discussion on whether this approach is acceptable

ericmj commented:
Does it need to be per query instead of global configuration?

talentdeficit commented:
it doesn't need to be per query, no. if this is an acceptable approach from an encoding/decoding perspective i can make it a per connection option also. i would like to retain the ability to determine it per query though as i use postgrex from both erlang and elixir in the same release and i would like to share the connection pool between them but also work with 'native' data types

ericmj commented:
Extensions are per connection so I don't think it makes sense to make this per query.

talentdeficit commented:
ok, that makes sense. i'll have to talk to @fishcakez as to the best way to proceed then as `db_connection` currently doesn't pass the connection opts through to queries

fishcakez commented:
@talentdeficit I think we would need to add `null` to the `Postgrex.Protocol` state inside `connect/1` because that is the only time postgrex can get the connection options. Then when we add the types table to the `Postgrex.Query` struct (inside `describe_recv`) we also add the `null` value to the query struct. This should mean the null value selection behaves like an extension and can fetch it when using the `DBConnection.Query` protocol.

If we are to set the the `null` value at query time using options I think it should occur in the describe callback of `Postgrex.Query` because that is when extension lookups occur. That means if using prepare/execute setting the `null` on prepare. From that perspective if you squint a bit it sort of makes sense. However I leave the decision up to @ericmj.

talentdeficit commented:
i've added per connection null mapping (via the `null` argument to `Postgrex.start_link`) and modified the query time mapping so that it works with prepare/execute as per @fishcakez suggestions. i added the options to the relevant documentation and there's example usage in `null_mapping_test.exs` in the test directory

ericmj commented:
This looks good to me. Waiting for @fishcakez to also sign off also

fishcakez commented on lib/postgrex.ex:
```diff
@@ -62,6 +62,8 @@ defmodule Postgrex do
     * `:pool` - The pool module to use, see `DBConnection`, it must be
     included with all requests if not the default (default:
     `DBConnection.Connection`);
+    * `:null` - The atom to use as a stand in for postgres' `NULL' in encoding
```
Should be markdown **`**NULL**`** not the erlang doc style **`**NULL**'**, same error in the other doc additions in this module.

fishcakez commented:
The implementation is great just a tiny error in the docs in 3 places.

talentdeficit commented:
fixed the incorrect markdown in the docs and a couple other minor doc issues also

fishcakez commented:
:+1:
